Core/Core/Use Cases/UploadManager.swift

```diff
@@ -215,32 +198,81 @@ open class UploadManager: NSObject, URLSessionDelegate, URLSessionTaskDelegate,
     }
 
     public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
-        Logger.shared.log()
         context.performAndWait {
             guard
                 let step = task.uploadStep,
                 let file = self.file(taskID: task.taskIdentifier)
             else { return }
+            let decoder = JSONDecoder()
+            decoder.dateDecodingStrategy = .iso8601
             switch step {
             case .upload:
-                if case let .submission(courseID, assignmentID, comment)? = file.context, error == nil {
-                    submit(file: file, courseID: courseID, assignmentID: assignmentID, comment: comment)
-                    return
+                // On device the upload might fail with a 201
+                // In this case, get the file using the url in the Location header
+                if error != nil,
+                    let response = task.response as? HTTPURLResponse,
+                    response.statusCode == 201,
+                    let location = response.allHeaderFields[HttpHeader.location] as? String,
+                    let url = URL(string: location),
+                    let user = file.user,
+                    let session = Keychain.entries.first(where: { user == $0 }) {
+                    do {
+                        var request = URLRequest(url: url)
+                        request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: HttpHeader.authorization)
+                        request.setValue("application/json+canvas-string-ids", forHTTPHeaderField: HttpHeader.accept)
+                        let task = backgroundSession.dataTask(with: request)
+                        task.uploadStep = .upload
+                        file.taskID = task.taskIdentifier
+                        try context.save()
+                        task.resume()
+                    } catch {
+                        complete(file: file, error: error)
+                    }
+                    break
+                }
+                guard let data = file.uploadData else {
+                    complete(file: file, error: error)
+                    break
+                }
+                do {
+                    let result = try decoder.decode(APIFile.self, from: data)
+                    file.update(fromAPIModel: result)
+                    try context.save()
+                    if case let .submission(courseID, assignmentID, comment)? = file.context, error == nil {
+                        submit(file: file, courseID: courseID, assignmentID: assignmentID, comment: comment)
```

Comment:
do you need to call `complete` in this case?

Reply:
No, `complete` will cause the file to disappear from the uploading screen. I want it to hang around until the submission request finishes just in case that request fails (or takes a long time).

Reply:
Good call out though.

Reply:
cool cool cool